src/Workspaces/Core/Portable/Utilities/BKTree.cs

```diff
@@ -0,0 +1,184 @@
+        // work on top of an ArraySlice (which needs a char[]).  The edit distance code also
+        // wants to work on top of raw char[]s (both for speed, and so it can pool arrays
+        // to prevent lots of garbage).  Because of that we just keep this as a char[].
+        private readonly char[] _allLowerCaseCharacters;
+        private readonly ImmutableArray<Node> _nodes;
+        private readonly ImmutableArray<Edge> _edges;
+
+        private BKTree(char[] allLowerCaseCharacters, ImmutableArray<Node> nodes, ImmutableArray<Edge> edges)
+        {
+            _allLowerCaseCharacters = allLowerCaseCharacters;
+            _nodes = nodes;
+            _edges = edges;
+        }
+
+        public static BKTree Create(params string[] values)
+        {
+            return Create((IEnumerable<string>)values);
+        }
+
+        public static BKTree Create(IEnumerable<string> values)
+        {
+            return new Builder(values).Create();
+        }
+
+        public IList<string> Find(string value, int? threshold = null)
+        {
+            if (_nodes.Length == 0)
+            {
+                return SpecializedCollections.EmptyList<string>();
+            }
+
+            var lowerCaseCharacters = ArrayPool<char>.GetArray(value.Length);
+            try
+            {
+                for (var i = 0; i < value.Length; i++)
+                {
+                    lowerCaseCharacters[i] = char.ToLower(value[i]);
+                }
+
+                threshold = threshold ?? WordSimilarityChecker.GetThreshold(value);
+                var result = new List<string>();
+                Lookup(_nodes[0], lowerCaseCharacters, value.Length, threshold.Value, result);
+                return result;
+            }
+            finally
+            {
+                ArrayPool<char>.ReleaseArray(lowerCaseCharacters);
+            }
+        }
+
+        private void Lookup(Node currentNode, char[] queryCharacters, int queryLength, int threshold, List<string> result)
+        {
+            // We always want to compute the real edit distance (ignoring any thresholds).  This is
+            // because we need that edit distance to appropriately determine which edges to walk 
+            // in the tree.
+            var characterSpan = currentNode.CharacterSpan;
+            var editDistance = EditDistance.GetEditDistance(
+                new ArraySlice<char>(_allLowerCaseCharacters, characterSpan),
+                new ArraySlice<char>(queryCharacters, 0, queryLength));
+
+            if (editDistance <= threshold)
+            {
+                // Found a match.
+                result.Add(new string(_allLowerCaseCharacters, characterSpan.Start, characterSpan.Length));
+            }
+
+            var min = editDistance - threshold;
+            var max = editDistance + threshold;
+
+            var firstEdgeIndex = currentNode.FirstEdgeIndex;
+            var lastEdgeIndex = firstEdgeIndex + currentNode.EdgeCount;
```

Comment:
The name "lastEdgeIndex" implies that it should be processed. I'm not sure what the right name of this thing would be, but you can -1 it and <= the loop and have it be correct and have the names make sense.

There were also a couple other "start/end" named things above that were slightly bothersome because they implied the "end" should be part of the set being evaluated, but those are more acceptable.